paddle/pten/common/scalar_array.h

```diff
@@ -0,0 +1,151 @@
+
+#pragma once
+
+#include "paddle/pten/api/ext/exception.h"
+#include "paddle/pten/api/include/tensor.h"
+
+namespace paddle {
+namespace experimental {
+
+template <typename T>
+class ScalarArrayBase {
+ public:
+  // Constructor support implicit
+  ScalarArrayBase() = default;
+
+  ScalarArrayBase(const std::vector<int64_t>& vec) : array_(vec) {}  // NOLINT
+
+  ScalarArrayBase(const int64_t* date_value, int64_t n) {
+    AssignData(date_value, n);
+  }
+
+  ScalarArrayBase(const int32_t* date_value, int64_t n) {
+    AssignData(date_value, n);
+  }
+
+  // The Tensor must have one dim
+  ScalarArrayBase(const T& tensor) {  // NOLINT
+    size_t n = tensor.numel();
+    array_.reserve(n);
+    switch (tensor.type()) {
+      case DataType::INT32:
+        AssignData(tensor.template data<int32_t>(), n);
+        break;
+      case DataType::INT64:
+        AssignData(tensor.template data<int64_t>(), n);
+        break;
+      default:
+        PD_THROW(
+            "Data type error. Currently, The data type of ScalarArrayBase "
+            "only supports Tensor with int32 and int64, "
+            "but now received `",
+            tensor.type(),
+            "`.");
+    }
+  }
+
+  // The Tensor in vec must have only one element
+  ScalarArrayBase(const std::vector<T>& tensor_list) {  // NOLINT
+    auto n = tensor_list.size();
+    array_.reserve(n);
+    if (!tensor_list.empty()) {
+      DataType data_type = tensor_list[0].dtype();
+      switch (data_type) {
+        case DataType::INT32: {
+          for (size_t i = 0; i < n; i++) {
+            PD_CHECK(tensor_list[i].dtype() == data_type,
+                     "The data_type of tensors in the list isn't consistent."
+                     "the first tensor is`",
+                     data_type,
+                     "` but `",
+                     i,
+                     "`th tensor is`",
+                     tensor_list[i].dtype(),
+                     "`.");
+            array_.push_back(*tensor_list[i].template data<int32_t>());
+          }
+          break;
+        }
+        case DataType::INT64: {
+          for (size_t i = 0; i < n; i++) {
+            PD_CHECK(tensor_list[i].dtype() == data_type,
+                     "The data_type of tensors in the list isn't consistent."
+                     "the first tensor is`",
+                     data_type,
+                     "` but `",
+                     i,
+                     "`th tensor is`",
+                     tensor_list[i].dtype(),
+                     "`.");
+            array_.push_back(*tensor_list[i].template data<int64_t>());
+          }
+          break;
+        }
+        default:
+          PD_THROW(
+              "Data type error. Currently, The data type of ScalarArrayBase "
+              "only supports Tensor with int32 and int64, "
+              "but now received `",
+              data_type,
+              "`.");
+      }
+    }
+  }
+
+  template <typename TT>
```

Comment:
TT is not a good name, maybe we can use InT, or OtherT?

Reply:
Done